doc/functions/library/test.py

```diff
@@ -0,0 +1,124 @@
+#!/usr/bin/env python3
+
+from pprint import pprint
+import xml.etree.ElementTree as ET
+import subprocess
+import json
+import os
+import glob
+import itertools
+import sys
+import shlex
+import argparse
+
+def debug(msg):
+    print(msg, file=sys.stderr)
+
+def find_examples(tree):
+    ns = {
+        'docbook': 'http://docbook.org/ns/docbook',
+    }
+
+    for example in tree.findall('.//docbook:example', ns):
+        xml_id = example.attrib['{http://www.w3.org/XML/1998/namespace}id']
+
+        title = example.find('docbook:title', ns)
+        if xml_id is None and title is None:
+            debug('Skipping <example> with no <title> or xml:id')
+            continue
+        elif xml_id is None:
+            debug('Skipping <example> with title "{}" because it has no xml:id'.format(title.text))
+            continue
+
+        code = example.find('.//docbook:programlisting[@role="input"]', ns)
+        if code is None:
+            debug('Skipping example {}: no <programlisting role="input">'.format(xml_id))
+            continue
+
+        expect = example.find('.//docbook:programlisting[@role="output"]', ns)
+        if expect is None:
+            debug('Skipping example {}: no <programlisting role="output">'.format(xml_id))
+            continue
+
+        yield { "id": xml_id, "code": code.text, "expect": expect.text }
+
+
+parser = argparse.ArgumentParser(description='Execute doc tests for the Nixpkgs manual..')
+default_lib_dir = os.path.realpath(os.path.dirname(os.path.dirname(os.path.realpath(__file__)) + "/../../../lib/"))
+parser.add_argument('--lib-dir', dest='library_dir', type=os.path.realpath,
+                    default=default_lib_dir,
+                    help='Source of the library to test against (default: {})'.format(default_lib_dir))
+parser.add_argument('ids', metavar='id', type=str, nargs='*',
+                    help='an example ID to test')
+
+args = parser.parse_args()
+
+doc_root = os.path.dirname(os.path.realpath(__file__))
+all_doc_files = glob.glob("{}/*.xml".format(doc_root))
+
+all_docs = [ET.parse(filename) for filename in all_doc_files]
+all_examples = itertools.chain(*[find_examples(doc) for doc in all_docs])
+
+if len(args.ids) > 1:
```

Comment:
if args.ids:

Reply:
I don't think so, this is >1 not >0

Reply:
... But should it be? seems like we drop a singleton list here? OK agreed with @FRidh then

Reply:
+1 this is a relic of when I didn't have proper argument parsing 😁 good catch.